testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/SSLMasterSlaveTwoWayTestCase.java

```diff
@@ -0,0 +1,128 @@
+import org.jboss.as.controller.client.ModelControllerClient;
+import org.jboss.as.controller.client.helpers.domain.DomainClient;
+import org.jboss.as.test.integration.domain.management.util.DomainLifecycleUtil;
+import org.jboss.as.test.integration.domain.management.util.DomainTestSupport;
+import org.jboss.as.test.integration.security.common.AbstractBaseSecurityRealmsServerSetupTask;
+import org.jboss.as.test.integration.security.common.SecurityTestConstants;
+import org.jboss.as.test.integration.security.common.config.realm.Authentication;
+import org.jboss.as.test.integration.security.common.config.realm.RealmKeystore;
+import org.jboss.as.test.integration.security.common.config.realm.SecurityRealm;
+import org.jboss.as.test.integration.security.common.config.realm.ServerIdentity;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Tests two way SSL secured communication between master and slave.
+ *
+ * @author Ondrej Kotek <okotek@redhat.com>
+ */
+public class SSLMasterSlaveTwoWayTestCase extends AbstractSSLMasterSlaveTestCase {
+
+    private static final File WORK_DIR = new File("target" + File.separatorChar +  "ssl-master-slave-2way-workdir");
+    public static final File SERVER_KEYSTORE_FILE = new File(WORK_DIR, SecurityTestConstants.SERVER_KEYSTORE);
+    public static final File SERVER_TRUSTSTORE_FILE = new File(WORK_DIR, SecurityTestConstants.SERVER_TRUSTSTORE);
+
+    private static final MasterManagementRealmSetup masterManagementRealmSetup = new MasterManagementRealmSetup();
+
+    private static DomainTestSupport testSupport;
+    private static DomainLifecycleUtil domainMasterLifecycleUtil;
+
+
+    @BeforeClass
+    public static void setupDomain() throws Exception {
+        keyMaterialSetup(WORK_DIR);
+
+        DomainTestSupport.Configuration configuration = DomainTestSupport.Configuration.create(
+                SSLMasterSlaveTwoWayTestCase.class.getSimpleName(), "domain-configs/domain-standard.xml",
+                "host-configs/host-master-ssl.xml", "host-configs/host-slave-ssl-2way.xml");
+
+        testSupport = DomainTestSupport.createAndStartSupport(configuration);
+        domainMasterLifecycleUtil = testSupport.getDomainMasterLifecycleUtil();
+
+        DomainClient masterClient = domainMasterLifecycleUtil.getDomainClient();
+
+        checkHostStatusOnMaster("slave");
+        masterManagementRealmSetup.setup(masterClient);
+        setMasterManagementNativeInterface(masterClient);
+        reloadMaster();
+        checkHostStatusOnMaster("master");
+    }
+
+    @AfterClass
+    public static void tearDownDomain() throws Exception {
+        setOriginMasterManagementNativeInterface();
+        reloadMaster();
+        checkHostStatusOnMaster("master");
+        masterManagementRealmSetup.tearDown(domainMasterLifecycleUtil.getDomainClient());
+
+        testSupport.stop();
+        testSupport = null;
+        domainMasterLifecycleUtil = null;
```

Comment:
These last 3 lines should be in a finally block. If there's a problem with this test, it must still shut down the processes or the rest of the testsuite will be screwed up.